File: cmd/mdtask/stats.go
package mdtask

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"github.com/tkancf/mdtask/internal/cli"
	"github.com/tkancf/mdtask/internal/task"
)

var statsCmd = &cobra.Command{
	Use:   "stats",
	Short: "Show task statistics",
	Long:  `Display statistics about your tasks including daily progress, completion rates, and status breakdown.`,
	RunE:  runStats,
}

var (
	statsDate   string
	statsWeek   bool
	statsMonth  bool
	statsSimple bool
)

func init() {
	rootCmd.AddCommand(statsCmd)
	statsCmd.Flags().StringVarP(&statsDate, "date", "d", "", "Show stats for specific date (YYYY-MM-DD)")
	statsCmd.Flags().BoolVarP(&statsWeek, "week", "w", false, "Show stats for current week")
	statsCmd.Flags().BoolVarP(&statsMonth, "month", "m", false, "Show stats for current month")
	statsCmd.Flags().BoolVarP(&statsSimple, "simple", "s", false, "Show simple output without graphics")
}

type TaskStats struct {
	Total       int `json:"total"`
	ByStatus    struct {
		TODO int `json:"todo"`
		WIP  int `json:"wip"`
		WAIT int `json:"wait"`
		SCHE int `json:"sche"`
		DONE int `json:"done"`
	} `json:"by_status"`
	Activity struct {
		Created   int `json:"created"`
		Updated   int `json:"updated"`
		Completed int `json:"completed"`
	} `json:"activity"`
	Deadlines struct {
		Overdue  int `json:"overdue"`
		Upcoming int `json:"upcoming"`
	} `json:"deadlines"`
	DateRange struct {
		Start string `json:"start"`
		End   string `json:"end"`
	} `json:"date_range"`
}

func runStats(cmd *cobra.Command, args []string) error {
	ctx, err := cli.LoadContext(cmd)
	if err != nil {
		return err
	}

	tasks, err := ctx.Repo.FindAll()
	if err != nil {
		return err
	}

	// Determine the date range
	var startDate, endDate time.Time
	now := time.Now()
	
	if statsDate != "" {
		// Specific date
		startDate, err = time.Parse("2006-01-02", statsDate)
		if err != nil {
			return fmt.Errorf("invalid date format: %w", err)
		}
		endDate = startDate.AddDate(0, 0, 1)
	} else if statsWeek {
		// Current week (Monday to Sunday)
		weekday := int(now.Weekday())
		if weekday == 0 {
			weekday = 7
		}
		startDate = now.AddDate(0, 0, -(weekday - 1))
		endDate = startDate.AddDate(0, 0, 7)
	} else if statsMonth {
		// Current month
		startDate = time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, now.Location())
		endDate = startDate.AddDate(0, 1, 0)
	} else {
		// Today (default)
		startDate = time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, now.Location())
		endDate = startDate.AddDate(0, 0, 1)
	}

	stats := calculateStats(tasks, startDate, endDate, now)
	
	if outputFormat == "json" {
		encoder := json.NewEncoder(os.Stdout)
		encoder.SetIndent("", "  ")
		return encoder.Encode(stats)
	}
	
	if statsSimple {
		displaySimpleStats(stats, startDate, endDate)
	} else {
		displayDetailedStats(stats, startDate, endDate, tasks)
	}

	return nil
}

func calculateStats(tasks []*task.Task, startDate, endDate, now time.Time) TaskStats {
	stats := TaskStats{}
	stats.DateRange.Start = startDate.Format("2006-01-02")
	stats.DateRange.End = endDate.AddDate(0, 0, -1).Format("2006-01-02")
	
	for _, t := range tasks {
		if t.IsArchived() {
			continue
		}
		
		stats.Total++
		
		// Count by status
		switch t.GetStatus() {
		case task.StatusTODO:
			stats.ByStatus.TODO++
		case task.StatusWIP:
			stats.ByStatus.WIP++
		case task.StatusWAIT:
			stats.ByStatus.WAIT++
		case task.StatusSCHE:
			stats.ByStatus.SCHE++
		case task.StatusDONE:
			stats.ByStatus.DONE++
		}
		
		// Check if created in date range
		if t.Created.After(startDate) && t.Created.Before(endDate) {
			stats.Activity.Created++
		}
		
		// Check if updated in date range
		if t.Updated.After(startDate) && t.Updated.Before(endDate) {
			stats.Activity.Updated++
		}
		
		// Check if completed in date range (assuming DONE tasks were completed when last updated)
		if t.GetStatus() == task.StatusDONE && t.Updated.After(startDate) && t.Updated.Before(endDate) {
			stats.Activity.Completed++
		}
		
		// Check deadlines
		if deadline := t.GetDeadline(); deadline != nil {
			if deadline.Before(now) {
				stats.Deadlines.Overdue++
			} else if deadline.Sub(now) < 7*24*time.Hour {
				stats.Deadlines.Upcoming++
			}
		}
	}
	
	return stats
}

func displaySimpleStats(stats TaskStats, startDate, endDate time.Time) {
	dateRange := formatDateRange(startDate, endDate)
	fmt.Printf("Task Statistics for %s\n", dateRange)
	fmt.Println(strings.Repeat("-", 40))
	fmt.Printf("Total Active Tasks: %d\n", stats.Total)
	fmt.Printf("  TODO: %d\n", stats.ByStatus.TODO)
	fmt.Printf("  WIP:  %d\n", stats.ByStatus.WIP)
	fmt.Printf("  WAIT: %d\n", stats.ByStatus.WAIT)
	fmt.Printf("  SCHE: %d\n", stats.ByStatus.SCHE)
	fmt.Printf("  DONE: %d\n", stats.ByStatus.DONE)
	fmt.Println()
	fmt.Printf("Created:   %d\n", stats.Activity.Created)
	fmt.Printf("Updated:   %d\n", stats.Activity.Updated)
	fmt.Printf("Completed: %d\n", stats.Activity.Completed)
	fmt.Println()
	fmt.Printf("Overdue Tasks:      %d\n", stats.Deadlines.Overdue)
	fmt.Printf("Upcoming Deadlines: %d\n", stats.Deadlines.Upcoming)
}

func displayDetailedStats(stats TaskStats, startDate, endDate time.Time, tasks []*task.Task) {
	dateRange := formatDateRange(startDate, endDate)
	
	// Header
	fmt.Printf("\n📊 Task Statistics for %s\n", dateRange)
	fmt.Println(strings.Repeat("═", 60))
	
	// Progress Summary
	fmt.Println("\n📈 Progress Summary")
	fmt.Println(strings.Repeat("─", 40))
	
	if stats.Activity.Created > 0 || stats.Activity.Completed > 0 {
		fmt.Printf("✨ Created:   %d new task(s)\n", stats.Activity.Created)
		fmt.Printf("✅ Completed: %d task(s)\n", stats.Activity.Completed)
		fmt.Printf("📝 Updated:   %d task(s)\n", stats.Activity.Updated)
	} else {
		fmt.Println("No activity in this period")
	}
	
	// Status Breakdown with progress bars
	fmt.Println("\n📋 Status Breakdown")
	fmt.Println(strings.Repeat("─", 40))
	
	if stats.Total > 0 {
		displayStatusBar("TODO", stats.ByStatus.TODO, stats.Total, "🔵")
		displayStatusBar("WIP ", stats.ByStatus.WIP, stats.Total, "🟡")
		displayStatusBar("WAIT", stats.ByStatus.WAIT, stats.Total, "⚪")
		displayStatusBar("SCHE", stats.ByStatus.SCHE, stats.Total, "🟣")
		displayStatusBar("DONE", stats.ByStatus.DONE, stats.Total, "🟢")
		
		fmt.Printf("\nTotal Active Tasks: %d\n", stats.Total)
		
		// Completion rate for the period
		if stats.Activity.Created > 0 {
			completionRate := float64(stats.Activity.Completed) / float64(stats.Activity.Created) * 100
			fmt.Printf("\n🎯 Daily Completion Rate: %.1f%% (%d/%d)\n", 
				completionRate, stats.Activity.Completed, stats.Activity.Created)
		}
	} else {
		fmt.Println("No active tasks")
	}
	
	// Alerts
	if stats.Deadlines.Overdue > 0 || stats.Deadlines.Upcoming > 0 {
		fmt.Println("\n⚠️  Alerts")
		fmt.Println(strings.Repeat("─", 40))
		
		if stats.Deadlines.Overdue > 0 {
			fmt.Printf("🚨 Overdue Tasks: %d\n", stats.Deadlines.Overdue)
			// List overdue tasks
			for _, t := range tasks {
				if !t.IsArchived() && t.GetDeadline() != nil && t.GetDeadline().Before(time.Now()) {
					fmt.Printf("   - %s (due: %s)\n", t.Title, t.GetDeadline().Format("2006-01-02"))
				}
			}
		}
		
		if stats.Deadlines.Upcoming > 0 {
			fmt.Printf("\n📅 Upcoming Deadlines (next 7 days): %d\n", stats.Deadlines.Upcoming)
		}
	}
	
	// Tasks in progress
	if stats.ByStatus.WIP > 0 {
		fmt.Println("\n🚧 Tasks in Progress")
		fmt.Println(strings.Repeat("─", 40))
		for _, t := range tasks {
			if !t.IsArchived() && t.GetStatus() == task.StatusWIP {
				fmt.Printf("- %s\n", t.Title)
			}
		}
	}
	
	fmt.Println()
}

func displayStatusBar(label string, count, total int, emoji string) {
	if total == 0 {
		return
	}
	
	percentage := float64(count) / float64(total) * 100
	barLength := 20
	filled := int(float64(barLength) * float64(count) / float64(total))
	
	bar := strings.Repeat("█", filled) + strings.Repeat("░", barLength-filled)
	
	fmt.Printf("%s %s [%s] %3d (%.1f%%)\n", emoji, label, bar, count, percentage)
}

func formatDateRange(startDate, endDate time.Time) string {
	duration := endDate.Sub(startDate)
	
	if duration.Hours() <= 24 {
		// Single day
		if startDate.Format("2006-01-02") == time.Now().Format("2006-01-02") {
			return "Today"
		}
		return startDate.Format("2006-01-02")
	} else if duration.Hours() <= 24*7 {
		// Week
		return fmt.Sprintf("Week of %s", startDate.Format("2006-01-02"))
	} else if duration.Hours() <= 24*31 {
		// Month
		return startDate.Format("January 2006")
	}
	
	return fmt.Sprintf("%s to %s", startDate.Format("2006-01-02"), endDate.AddDate(0, 0, -1).Format("2006-01-02"))
}

File: nvim-mdtask/lua/mdtask/stats.lua
local M = {}

-- Calculate task statistics
function M.calculate_stats(tasks)
  local stats = {
    total = 0,
    by_status = {
      TODO = 0,
      WIP = 0,
      WAIT = 0,
      SCHE = 0,
      DONE = 0,
    },
    overdue = 0,
    due_today = 0,
    due_this_week = 0,
    has_deadline = 0,
    no_deadline = 0,
    completion_rate = 0,
    created_today = 0,
    created_this_week = 0,
    created_this_month = 0,
    completed_today = 0,
    completed_this_week = 0,
    completed_this_month = 0,
  }
  
  -- Get current date info
  local now = os.time()
  local today = os.date("*t", now)
  today.hour = 0
  today.min = 0
  today.sec = 0
  local today_start = os.time(today)
  
  -- Calculate week start (Sunday)
  local days_since_sunday = today.wday - 1
  local week_start = today_start - (days_since_sunday * 24 * 60 * 60)
  
  -- Calculate month start
  local month_start = os.time({
    year = today.year,
    month = today.month,
    day = 1,
    hour = 0,
    min = 0,
    sec = 0
  })
  
  for _, task in ipairs(tasks) do
    stats.total = stats.total + 1
    
    -- Count by status
    local status = task.status or 'TODO'
    if stats.by_status[status] then
      stats.by_status[status] = stats.by_status[status] + 1
    end
    
    -- Check deadline
    if task.deadline then
      stats.has_deadline = stats.has_deadline + 1
      
      local deadline_time = vim.fn.strptime('%Y-%m-%dT%H:%M:%SZ', task.deadline)
      if deadline_time then
        if deadline_time < now then
          stats.overdue = stats.overdue + 1
        elseif deadline_time < today_start + (24 * 60 * 60) then
          stats.due_today = stats.due_today + 1
        elseif deadline_time < week_start + (7 * 24 * 60 * 60) then
          stats.due_this_week = stats.due_this_week + 1
        end
      end
    else
      stats.no_deadline = stats.no_deadline + 1
    end
    
    -- Check created date
    if task.created then
      local created_time = vim.fn.strptime('%Y-%m-%dT%H:%M:%SZ', task.created)
      if created_time then
        if created_time >= today_start then
          stats.created_today = stats.created_today + 1
        end
        if created_time >= week_start then
          stats.created_this_week = stats.created_this_week + 1
        end
        if created_time >= month_start then
          stats.created_this_month = stats.created_this_month + 1
        end
      end
    end
    
    -- Check completed date (if task is DONE)
    if status == 'DONE' and task.updated then
      local updated_time = vim.fn.strptime('%Y-%m-%dT%H:%M:%SZ', task.updated)
      if updated_time then
        if updated_time >= today_start then
          stats.completed_today = stats.completed_today + 1
        end
        if updated_time >= week_start then
          stats.completed_this_week = stats.completed_this_week + 1
        end
        if updated_time >= month_start then
          stats.completed_this_month = stats.completed_this_month + 1
        end
      end
    end
  end
  
  -- Calculate completion rate
  if stats.total > 0 then
    stats.completion_rate = math.floor((stats.by_status.DONE / stats.total) * 100)
  end
  
  return stats
end

-- Format stats for display
function M.format_stats(stats)
  local lines = {}
  
  -- Header
  table.insert(lines, "Task Statistics")
  table.insert(lines, string.rep("═", 40))
  table.insert(lines, "")
  
  -- Overall stats
  table.insert(lines, "Overall:")
  table.insert(lines, string.format("  Total Tasks: %d", stats.total))
  table.insert(lines, string.format("  Completion Rate: %d%%", stats.completion_rate))
  table.insert(lines, "")
  
  -- Status breakdown
  table.insert(lines, "By Status:")
  table.insert(lines, string.format("  TODO: %d", stats.by_status.TODO))
  table.insert(lines, string.format("  WIP:  %d", stats.by_status.WIP))
  table.insert(lines, string.format("  WAIT: %d", stats.by_status.WAIT))
  table.insert(lines, string.format("  SCHE: %d", stats.by_status.SCHE))
  table.insert(lines, string.format("  DONE: %d", stats.by_status.DONE))
  table.insert(lines, "")
  
  -- Deadline stats
  table.insert(lines, "Deadlines:")
  table.insert(lines, string.format("  Overdue: %d", stats.overdue))
  table.insert(lines, string.format("  Due Today: %d", stats.due_today))
  table.insert(lines, string.format("  Due This Week: %d", stats.due_this_week))
  table.insert(lines, string.format("  Has Deadline: %d", stats.has_deadline))
  table.insert(lines, string.format("  No Deadline: %d", stats.no_deadline))
  table.insert(lines, "")
  
  -- Activity stats
  table.insert(lines, "Activity:")
  table.insert(lines, string.format("  Created Today: %d", stats.created_today))
  table.insert(lines, string.format("  Created This Week: %d", stats.created_this_week))
  table.insert(lines, string.format("  Created This Month: %d", stats.created_this_month))
  table.insert(lines, string.format("  Completed Today: %d", stats.completed_today))
  table.insert(lines, string.format("  Completed This Week: %d", stats.completed_this_week))
  table.insert(lines, string.format("  Completed This Month: %d", stats.completed_this_month))
  table.insert(lines, "")
  
  -- Visual progress bar for completion rate
  table.insert(lines, "Progress:")
  local bar_width = 30
  local filled = math.floor((stats.completion_rate / 100) * bar_width)
  local empty = bar_width - filled
  local bar = "  [" .. string.rep("█", filled) .. string.rep("░", empty) .. "] " .. stats.completion_rate .. "%"
  table.insert(lines, bar)
  
  return lines
end

return M